runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -395,12 +402,21 @@ public static StreamingDataflowWorker fromOptions(DataflowWorkerHarnessOptions o
         workFailureProcessor,
         streamingCounters,
         memoryMonitor,
-        maxWorkItemCommitBytes,
+        operationalLimits,
         configFetcherComputationStateCacheAndWindmillClient.windmillStreamFactory(),
         executorSupplier,
         stageInfo);
   }
 
+  public static class OperationalLimits {
+    // Maximum size of a commit from a single work item.
+    public long maxWorkItemCommitBytes;
```

Comment:
I think that this class should be immutable (ie all final members), and then you make new instances of OperationLimits whenever you want to modify it.

As is, there are data races where the object referred to is read in the atomic ref (atomically) but then examining it's members is not atomic and simultaneously the background thread could modify the same object.

Reply:
Could use a AutoBuilder to simplify creating the immutable object

```
class OperationalLimits {
  final long maxWorkItemCommitBytes;
  OperationalLimits(long maxWorkItemCommitBytes, ...) { ...}

  @AutoBuilder(ofClass = OperationalLimits.class)
  public interface Builder {
    Builder setMaxWorkItemCommitBytes(long bytes);
    ...
    OperationalLimits build();
  }

  static Builder builder() {
    return AutoBuilder_Builder().setMaxWorkItemCommitBytes(default).set....();
  }
}
```

Reply:
Changed to immutable + AutoBuilder.